docs/topics/genericgateway.md

```diff
@@ -0,0 +1,65 @@
+# Generic Gateway
+
+Brigade contains a Generic Gateway and can be used to accept requests from other platforms or systems. Generic Gateway is a separate component in the Brigade system, like Github and Container Registry (CR) Gateways.
+
+Generic Gateway is _not enabled by default_.
+
+## Intro to Generic Gateway 
+
+Generic Gateway can optionally be activated to accept `POST` webhook requests at `/webhook/:projectID/:secret` path. When this endpoint is called, Brigade will respond by creating a Build with a `webhook` event. This provides Brigade developers with the ability to trigger scripts based on messages received from any platform that can send a POST HTTP request.
+
+## Generic Gateway
+
+Generic Gateway support is disabled by default, but can easily be turned on during installation or upgrade of Brigade:
+
+```
+$ helm install -n brigade brigade/brigade --set genericGateway.enabled=true
+```
+
+This will enable the Generic Gateway Deployment/Services/RBAC permissions. However, be aware that Generic Gateway is not exposed outside the cluster. In case you are certain from a security perspective that you want to expose the Generic Gateway via a [Kubernetes LoadBalancer Service](https://kubernetes.io/docs/concepts/services-networking/#loadbalancer), you can install Brigade via the following command:
+
+```
+$ helm install -n brigade brigade/brigade --set genericGateway.enabled=true,genericGateway.service.type=LoadBalancer
+```
+
+Alternatively, for enhanced security, you can install an SSL proxy (like `cert-manager`) and direct it to the Generic Gateway Service.
+
+## Using the Generic Gateway
+
+As mentioned, Generic Gateway accepts POST requests at `/webhook/:projectID/:secret` endpoint. These requests should also carry a JSON payload.
+
+- `projectID` is the Brigade Project ID
+- `secret` is a custom secret for this specific project's Generic Gateway webhook support. You can think of it as a simple authentication mechanism for this Project's Generic Gateway webhook support. Every Project has its own unique secret.
+
+When you create a new Brigade Project via Brig CLI, you can optionally create such a secret by adding a secret named `genericGatewaySecret`, containing your desired value. Alternatively, if you do you use the Advanced Options during `brig project create`, Brig will generate and output one for you.
```

Comment:
This is a bit confusing.
Would it make sense to directly add a prompt for the generic gateway secret in the advanced options?

Reply:
Also, I think there's a bug somewhere in setting the generic gateway secret - consider the following scenario: I create a new project and choose advanced options:

```
radu:brigade$ ./bin/brig project create --namespace brigade
? Project Name abcdef
? Full repository name github.com/abcdef
? Clone URL (https://github.com/your/repo.git) https://github.com/abcdef.git
? Add secrets? No
Auto-generated a Shared Secret: "m2XClW0LTMfwgoPWLARw5trT"
? Configure GitHub Access? No
? Configure advanced options Yes
? Custom VCS sidecar (enter 'NONE' for no sidecar) deis/git-sidecar:latest
? Build storage size 
? SecretKeyRef usage No
? Build storage class standard
? Job cache storage class standard
? Worker image registry or DockerHub org 
? Worker image name 
? Custom worker image tag 
? Worker image pull policy IfNotPresent
? Worker command yarn -s start
? Initialize Git submodules No
? Allow host mounts No
? Allow privileged jobs No
? Image pull secrets 
? Default script ConfigMap name 
? brigade.js file path relative to the repository root 
? Upload a default brigade.js script 
Auto-generated Generic Gateway Secret (for Brigade API Server): A1m9n
Project ID: brigade-bef57ec7f53a6d40beb640a780a639c83bc29ac8a9816f1fc6c5c6
radu:brigade$ brig project get brigade-bef57ec7f53a6d40beb640a780a639c83bc29ac8a9816f1fc6c5c6 --namespace brigade
{
  "metadata": {
    "name": "brigade-bef57ec7f53a6d40beb640a780a639c83bc29ac8a9816f1fc6c5c6",
    "creationTimestamp": null,
    "labels": {
      "app": "brigade",
      "component": "project",
      "heritage": "brigade"
    },
    "annotations": {
      "projectName": "abcdef"
    }
  },
  "stringData": {
    "allowHostMounts": "false",
    "allowPrivilegedJobs": "false",
    "brigadejsPath": "",
    "buildStorageSize": "50Mi",
    "cloneURL": "https://github.com/abcdef.git",
    "defaultScript": "",
    "defaultScriptName": "",
    "github.baseURL": "",
    "github.token": "",
    "github.uploadURL": "",
    "imagePullSecrets": "",
    "initGitSubmodules": "false",
    "kubernetes.allowSecretKeyRef": "false",
    "kubernetes.buildStorageClass": "standard",
    "kubernetes.cacheStorageClass": "standard",
    "namespace": "brigade",
    "repository": "github.com/abcdef",
    "secrets": "{}",
    "sharedSecret": "m2XClW0LTMfwgoPWLARw5trT",
    "sshKey": "",
    "vcsSidecar": "deis/git-sidecar:latest",
    "worker.name": "",
    "worker.pullPolicy": "IfNotPresent",
    "worker.registry": "",
    "worker.tag": "",
    "workerCommand": "yarn -s start"
  },
  "type": "brigade.sh/project"
}
radu:brigade$ brig version
v0.19.0-82-g3ae57cc
```

After creating the project, it tells me `Auto-generated Generic Gateway Secret (for Brigade API Server): A1m9n`. But that is nowhere to be found in the project.

Reply:
It is also unclear from this doc how one would create the secret, and how it is used. I added a new secret named `genericGatewaySecret` (which I think is what the gateway is looking for?):

```
radu:brigade$ brig project create --namespace brigade
? Project Name manually-added-secreat
? Full repository name github.com/manually-added-secreat
? Clone URL (https://github.com/your/repo.git) https://github.com/manually-added-secreat.git
? Add secrets? Yes
?       Secret 1 genericGatewaySecret
?       Value testing
? ===> Add another? No
Auto-generated a Shared Secret: "3r2RaJ8AsEtcKz18nPKHbE8X"
? Configure GitHub Access? No
? Configure advanced options No
Project ID: brigade-6f06bff6a9d15af9ac255804c4904c0a0cb97bc5e73f1f67c2250f

radu:brigade$ brig project get --namespace brigade brigade-6f06bff6a9d15af9ac255804c4904c0a0cb97bc5e73f1f67c2250f
{
  "metadata": {
    "name": "brigade-6f06bff6a9d15af9ac255804c4904c0a0cb97bc5e73f1f67c2250f",
    "creationTimestamp": null,
    "labels": {
      "app": "brigade",
      "component": "project",
      "heritage": "brigade"
    },
    "annotations": {
      "projectName": "manually-added-secreat"
    }
  },
  "stringData": {
    "allowHostMounts": "false",
    "allowPrivilegedJobs": "false",
    "brigadejsPath": "",
    "buildStorageSize": "50Mi",
    "cloneURL": "https://github.com/manually-added-secreat.git",
    "defaultScript": "",
    "defaultScriptName": "",
    "github.baseURL": "",
    "github.token": "",
    "github.uploadURL": "",
    "imagePullSecrets": "",
    "initGitSubmodules": "false",
    "kubernetes.allowSecretKeyRef": "false",
    "kubernetes.buildStorageClass": "",
    "kubernetes.cacheStorageClass": "",
    "namespace": "brigade",
    "repository": "github.com/manually-added-secreat",
    "secrets": "{\"genericGatewaySecret\":\"testing\"}",
    "sharedSecret": "3r2RaJ8AsEtcKz18nPKHbE8X",
    "sshKey": "",
    "vcsSidecar": "deis/git-sidecar:latest",
    "worker.name": "",
    "worker.pullPolicy": "IfNotPresent",
    "worker.registry": "",
    "worker.tag": "",
    "workerCommand": "yarn -s start"
  },
  "type": "brigade.sh/project"
}
```

That is added in the project as  ` "secrets": "{\"genericGatewaySecret\":\"testing\"}"`.

Now I'm trying to use the gateway:

```
$ curl http://localhost:8000/webhook/brigade-6f06bff6a9d15af9ac255804c4904c0a0cb97bc5e73f1f67c2250f/testing -X POST -H "Content-Type: application/json" -d '{"commit":"foo"}'

{"status":"secret is empty, please insert a secret in the URL and try again"}
```

And I suspect it's because in the project this secret is set as `sv.String("genericGatewaySecret")`, when it should probably be set as `proj.Secrets["genericGatewaySecret"]`.

Am I missing something here?

Reply:
Indeed, it was really confusing. Please take a look now, it should be good. I chose to not generate anything if user does not get into Advanced Options during `brig project create`, since they might not want to use the Generic Gateway, so it wouldn't be worth generating anything. Hope it's now clear in the docs as well.